### PR TITLE
Merge HSM branch with master to have a single executable

### PR DESCRIPTION
**Pull request by MarioDfinity**

### CODE REVIEW

**Review by jean-dfinity — APPROVED** [review]

**Review by jean-dfinity — DISMISSED** [review]

File: src/lib/mod.rs [hunk 1]
```diff
@@ -122,35 +163,57 @@ pub fn get_agent(pem: &str) -> AnyhowResult<Agent> {
...
+        AuthInfo::PemFile(pem) => match Secp256k1Identity::from_pem(pem.as_bytes()) {
+            Ok(id) => Ok(Box::new(id) as _),
+            Err(_) => match BasicIdentity::from_pem(pem.as_bytes()) {
+                Ok(id) => Ok(Box::new(id) as _),
+                Err(_) => Err(anyhow!("Couldn't load identity from PEM file")),
+            },
+        },
+        AuthInfo::NitroHsm(info) => {
+            let user_set_pin = { info.pin.borrow().clone() };
+            let pin_fn = || match user_set_pin {
```
Comment (roman-kashitsyn) [inline src/lib/mod.rs:201]: I don't think it works as was intended.
The original code requested the pin once and cached it. The new code will request the pin on each invocation of `pin_fn`.

File: src/lib/mod.rs [hunk 2]
```diff
@@ -122,35 +163,57 @@ pub fn get_agent(pem: &str) -> AnyhowResult<Agent> {
...
+                        *info.pin.borrow_mut() = Some(pin.clone());
+                        Ok(pin)
+                    }
+                    Err(e) => Err(e),
+                },
+                Some(pin) => Ok(pin),
+            };
+            HardwareIdentity::new(&info.libpath, info.slot, &info.ident, pin_fn)
+                .context("Unable to use your hardware key")
+                .map(|i| Box::new(i) as _)
```
Comment (roman-kashitsyn) [inline src/lib/mod.rs:215]: I'm not sure I understand why we need this box.

File: src/lib/mod.rs [hunk 3]
```diff
@@ -122,35 +163,57 @@ pub fn get_agent(pem: &str) -> AnyhowResult<Agent> {
...
-            Err(_) => {
-                eprintln!("Couldn't load identity from PEM file");
-                std::process::exit(1);
-            }
-        },
+fn ask_nitrohsm_pin_via_tty() -> Result<String, String> {
+    rpassword::read_password_from_tty(Some("NitroHSM PIN: "))
+        .context("Cannot read NitroHSM PIN from tty")
+        // TODO: better error string
+        .map_err(|e| format!("{}", e))
```
Comment (roman-kashitsyn) [inline src/lib/mod.rs:177]: nit:
```suggestion
        .map_err(|e| e.to_string())
```

File: src/commands/update_node_provider.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    auth: &AuthInfo,
+    opts: UpdateNodeProviderOpts,
+) -> AnyhowResult<Vec<IngressWithRequestId>> {
+    let reward_account = ledger_canister::AccountIdentifier::from_hex(&opts.reward_account)
+        .map_err(|e| {
+            anyhow::Error::msg(format!(
+                "Account {} is not valid address, {}",
+                &opts.reward_account, e,
+            ))
+        })?;
```
Comment (roman-kashitsyn) [inline src/commands/update_node_provider.rs:35]: Let's use anyhow features to keep errors properly nested
```suggestion
    let reward_account = ledger_canister::AccountIdentifier::from_hex(&opts.reward_account)
        .with_context(|| format!("Account {} is not valid address", &opts.reward_account))?;
```

**Review by roman-kashitsyn — DISMISSED** [review]
LGTM, left a few minor nits.

File: src/lib/mod.rs (hunk 1 above)
Comment (MarioDfinity) [reply src/lib/mod.rs]: Fixed with 53025ce.

File: src/lib/mod.rs (hunk 2 above)
Comment (MarioDfinity) [reply src/lib/mod.rs:217]: The returning type is `AnyhowResult<dyn Identity>` and `dyn Identity` doesn't have size.

File: src/lib/mod.rs [hunk 5]
```diff
@@ -122,35 +163,59 @@ pub fn get_agent(pem: &str) -> AnyhowResult<Agent> {
...
+                            Ok(pin)
+                        }
+                        Err(e) => Err(e),
+                    },
+                    Some(pin) => Ok(pin),
+                }
+            };
+            HardwareIdentity::new(&info.libpath, info.slot, &info.ident, pin_fn)
+                .context("Unable to use your hardware key")
+                .map(|i| Box::new(i) as _)
```
Comment (roman-kashitsyn) [inline src/lib/mod.rs:217]: Ah, I see, I think this way will be more readable:

```suggestion
            Box::new(HardwareIdentity::new(&info.libpath, info.slot, &info.ident, pin_fn)
                .context("Unable to use your hardware key")?)
```

**Review by roman-kashitsyn — DISMISSED** [review]

**Review by roman-kashitsyn — APPROVED** [review]

**Pull request merged**